src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,79 @@
+package com.example.onlinebookstore.service.impl;
+
+import com.example.onlinebookstore.dto.shoppingcart.CartItemDto;
+import com.example.onlinebookstore.dto.shoppingcart.ShoppingCartResponseDto;
+import com.example.onlinebookstore.exceptions.EntityNotFoundException;
+import com.example.onlinebookstore.mapper.ShoppingCartMapper;
+import com.example.onlinebookstore.model.Book;
+import com.example.onlinebookstore.model.CartItem;
+import com.example.onlinebookstore.model.ShoppingCart;
+import com.example.onlinebookstore.model.User;
+import com.example.onlinebookstore.repository.BookRepository;
+import com.example.onlinebookstore.repository.CartItemRepository;
+import com.example.onlinebookstore.repository.ShoppingCartRepository;
+import com.example.onlinebookstore.repository.UserRepository;
+import com.example.onlinebookstore.service.ShoppingCartService;
+import jakarta.transaction.Transactional;
+import java.util.HashSet;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final UserRepository userRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+
+    @Transactional
+    @Override
+    public ShoppingCartResponseDto getShoppingCart() {
+        return shoppingCartMapper.toResponseCart(findCart());
+    }
+
+    @Transactional
+    @Override
+    public ShoppingCartResponseDto addCartItem(CartItemDto cartItemDto) {
+        Book book = bookRepository.findById(cartItemDto.bookId()).orElseThrow(
+                () -> new EntityNotFoundException("Non existent book id"));
+        ShoppingCart cart = findCart();
+        CartItem cartItem = new CartItem();
+        cartItem.setBook(book);
+        cartItem.setQuantity(cartItemDto.quantity());
+        cartItem.setShoppingCart(cart);
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toResponseCart(cart);
+    }
+
+    @Override
+    public void deleteCartItem(Long id) {
+        cartItemRepository.deleteById(id);
+    }
+
+    @Transactional
+    @Override
+    public ShoppingCartResponseDto updateCartItem(Long quantity, Long id) {
+        ShoppingCart shoppingCartResponse = findCart();
+        CartItem cartItem = cartItemRepository
+                .findByIdAndShoppingCartId(id, shoppingCartResponse.getId())
+                .orElseThrow(() -> new EntityNotFoundException("Non existent cart item id"));
+        cartItem.setQuantity(quantity);
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toResponseCart(cartItem.getShoppingCart());
```

Comment:
pls make sure that your change is present when you do `cartItem.getShoppingCart()` 

since transaction is not commited yet, there's a risk of getting a shopping cart, where that particular cart item has previous quantity value

Reply:
it works